File: app/src/main/java/de/xikolo/controllers/login/LoginFragment.java
package de.xikolo.controllers.login;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.TextInputEditText;
import android.support.v4.app.NavUtils;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.EditorInfo;
import android.view.inputmethod.InputMethodManager;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;

import com.yatatsu.autobundle.AutoBundleField;

import butterknife.BindView;
import butterknife.OnClick;
import butterknife.OnEditorAction;
import de.xikolo.R;
import de.xikolo.config.Config;
import de.xikolo.config.GlideApp;
import de.xikolo.controllers.base.BasePresenterFragment;
import de.xikolo.controllers.dialogs.ProgressDialogIndeterminate;
import de.xikolo.controllers.dialogs.ProgressDialogIndeterminateAutoBundle;
import de.xikolo.managers.UserManager;
import de.xikolo.presenters.base.PresenterFactory;
import de.xikolo.presenters.login.LoginPresenter;
import de.xikolo.presenters.login.LoginPresenterFactory;
import de.xikolo.presenters.login.LoginView;
import de.xikolo.utils.ToastUtil;

public class LoginFragment extends BasePresenterFragment<LoginPresenter, LoginView> implements LoginView {

    public static final String TAG = LoginFragment.class.getSimpleName();

    @AutoBundleField(required = false) String token;

    @BindView(R.id.toolbar) Toolbar tb;

    @BindView(R.id.editEmail) TextInputEditText editTextEmail;
    @BindView(R.id.editPassword) TextInputEditText editTextPassword;
    @BindView(R.id.top_image) ImageView topImage;
    @BindView(R.id.text_credentials) TextView textCredentials;

    @BindView(R.id.btnSSO) Button buttonSSO;
    @BindView(R.id.ssoContainer) View containerSSO;

    private ProgressDialogIndeterminate progressDialog;

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_login, container, false);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        AppCompatActivity activity = (AppCompatActivity) getActivity();
        activity.setSupportActionBar(tb);
        ActionBar actionBar = activity.getSupportActionBar();
        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true);
            actionBar.setHomeButtonEnabled(true);
        }

        textCredentials.setText(String.format(getString(R.string.login_with_credentials), Config.HOST));

        GlideApp.with(this)
                .load(R.drawable.login_header_v2)
                .dontAnimate()
                .noPlaceholders()
                .fitCenter()
                .into(topImage);
    }

    @Override
    protected void onPresenterCreatedOrRestored(@NonNull LoginPresenter presenter) {
        if (token != null) {
            presenter.externalLoginCallback(token);
        }
    }

    @Override
    public void onResume() {
        super.onResume();

        if (UserManager.isAuthorized()) {
            getActivity().finish();
        }
    }

    @SuppressWarnings("unused")
    @OnEditorAction(R.id.editPassword)
    protected boolean editPasswordAction(TextView v, int actionId) {
        boolean handled = false;
        if (actionId == EditorInfo.IME_ACTION_DONE) {
            login(v);
            handled = true;
        }
        return handled;
    }

    @SuppressWarnings("unused")
    @OnClick(R.id.btnLogin)
    protected void loginClicked(View view) {
        login(view);
    }

    @SuppressWarnings("unused")
    @OnClick(R.id.btnNew)
    protected void newAccountClicked(View view) {
        hideKeyboard(view);
        startUrlIntent(Config.HOST_URL + Config.ACCOUNT + Config.NEW);
    }

    @SuppressWarnings("unused")
    @OnClick(R.id.textForgotPw)
    protected void forgotPasswordClicked(View view) {
        hideKeyboard(view);
        startUrlIntent(Config.HOST_URL + Config.ACCOUNT + Config.RESET);
    }

    @SuppressWarnings("unused")
    @OnClick(R.id.btnSSO)
    protected void ssoClicked(View view) {
        hideKeyboard(view);
        presenter.onSSOClicked();
    }

    @Override
    public void showSSOView() {
        containerSSO.setVisibility(View.VISIBLE);
    }

    private void login(View view) {
        hideKeyboard(view);
        String email = editTextEmail.getText().toString().trim();
        String password = editTextPassword.getText().toString();
        if (isEmailValid(email)) {
            if (!password.equals("")) {
                presenter.login(email, password);
            } else {
                editTextPassword.setError(getString(R.string.error_password));
            }
        } else {
            editTextEmail.setError(getString(R.string.error_email));
        }
    }

    @Override
    public void startSSOLogin(String strategy) {
        Intent intent = SsoLoginActivityAutoBundle.builder(
                Config.HOST_URL + "?in_app=true&redirect_to=" + strategy,
                getString(R.string.login_sso)
        ).build(getActivity());
        startActivity(intent);
    }

    @Override
    public void showProgressDialog() {
        if (progressDialog == null) {
            progressDialog = ProgressDialogIndeterminateAutoBundle.builder().build();
        }
        progressDialog.show(getChildFragmentManager(), ProgressDialogIndeterminate.TAG);
    }

    @Override
    public void hideProgressDialog() {
        if (progressDialog != null) {
            progressDialog.dismiss();
        }
    }

    @Override
    public void showNoNetworkToast() {
        ToastUtil.show(R.string.toast_no_network);
    }

    @Override
    public void showLoginFailedToast() {
        ToastUtil.show(R.string.toast_log_in_failed);
    }

    @Override
    public void finishActivity() {
        getActivity().finish();
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();
        if (id == android.R.id.home) {
            NavUtils.navigateUpFromSameTask(getActivity());
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    private void hideKeyboard(View view) {
        InputMethodManager imm = (InputMethodManager) getActivity().getSystemService(Context.INPUT_METHOD_SERVICE);
        imm.hideSoftInputFromWindow(view.getWindowToken(), 0);
    }

    private boolean isEmailValid(CharSequence email) {
        return android.util.Patterns.EMAIL_ADDRESS.matcher(email).matches();
    }

    private void startUrlIntent(String url) {
        Intent i = new Intent(Intent.ACTION_VIEW);
        i.setData(Uri.parse(url));
        startActivity(i);
    }

    @NonNull
    @Override
    protected PresenterFactory<LoginPresenter> getPresenterFactory() {
        return new LoginPresenterFactory();
    }

}

File: app/src/main/java/de/xikolo/controllers/login/LoginFragment.kt
package de.xikolo.controllers.login

import android.content.Context
import android.os.Bundle
import android.view.MenuItem
import android.view.View
import android.view.inputmethod.EditorInfo
import android.view.inputmethod.InputMethodManager
import android.widget.Button
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.core.app.NavUtils
import androidx.lifecycle.Observer
import butterknife.BindView
import com.google.android.material.textfield.TextInputEditText
import com.yatatsu.autobundle.AutoBundleField
import de.xikolo.App
import de.xikolo.R
import de.xikolo.config.Config
import de.xikolo.config.Feature
import de.xikolo.controllers.base.ViewModelFragment
import de.xikolo.controllers.helper.LoginHelper
import de.xikolo.managers.UserManager
import de.xikolo.network.jobs.base.NetworkCode
import de.xikolo.network.jobs.base.NetworkState
import de.xikolo.storages.UserStorage
import de.xikolo.utils.extensions.getString
import de.xikolo.utils.extensions.openUrl
import de.xikolo.utils.extensions.showToast
import de.xikolo.viewmodels.login.LoginViewModel

class LoginFragment : ViewModelFragment<LoginViewModel>() {

    companion object {
        val TAG: String = LoginFragment::class.java.simpleName
    }

    @AutoBundleField(required = false)
    var token: String? = null

    @BindView(R.id.toolbar)
    lateinit var toolbar: Toolbar

    @BindView(R.id.editEmail)
    lateinit var editTextEmail: TextInputEditText

    @BindView(R.id.editPassword)
    lateinit var editTextPassword: TextInputEditText

    @BindView(R.id.text_credentials)
    lateinit var textCredentials: TextView

    @BindView(R.id.textForgotPw)
    lateinit var textForgotPassword: TextView

    @BindView(R.id.btnSSO)
    lateinit var buttonSSO: Button

    @BindView(R.id.btnLogin)
    lateinit var buttonLogin: Button

    @BindView(R.id.btnNew)
    lateinit var buttonNew: Button

    @BindView(R.id.ssoContainer)
    lateinit var containerSSO: View

    override val layoutResource = R.layout.fragment_login

    override fun createViewModel(): LoginViewModel {
        return LoginViewModel()
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        (activity as? AppCompatActivity)?.apply {
            setSupportActionBar(toolbar)
            supportActionBar?.apply {
                setDisplayHomeAsUpEnabled(true)
                setHomeButtonEnabled(true)
            }
        }

        textCredentials.text = String.format(getString(R.string.login_with_credentials), Config.HOST)

        editTextPassword.setOnEditorActionListener { v, actionId, _ ->
            if (actionId == EditorInfo.IME_ACTION_DONE) {
                login(v)
                true
            } else {
                false
            }
        }

        buttonLogin.setOnClickListener {
            login(it)
        }

        buttonNew.setOnClickListener {
            hideKeyboard(view)
            if (activity?.openUrl(Config.HOST_URL + Config.ACCOUNT + Config.NEW) != true) {
                showToast(R.string.error_plain)
            }
        }


        if (Feature.enabled("sso_provider")) {
            containerSSO.visibility = View.VISIBLE
            buttonSSO.setOnClickListener {
                hideKeyboard(view)
                startSSOLogin()
            }

            if (Feature.enabled("disabled_registration")) {
                buttonNew.visibility = View.GONE
            }
        }

        textForgotPassword.setOnClickListener {
            hideKeyboard(view)
            if (activity?.openUrl(Config.HOST_URL + Config.ACCOUNT + Config.RESET) != true) {
                showToast(R.string.error_plain)
            }
        }

        viewModel.loginNetworkState
            .observeForever(
                object : Observer<NetworkState> {
                    override fun onChanged(it: NetworkState) {
                        when (it.code) {
                            NetworkCode.SUCCESS -> {
                                viewModel.loginNetworkState.removeObserver(this)

                                viewModel.requestUserWithProfile()
                            }
                            else                -> handleCode(it.code)
                        }
                    }
                }
            )

        viewModel.profileNetworkState
            .observeForever(
                object : Observer<NetworkState> {
                    override fun onChanged(it: NetworkState) {
                        when (it.code) {
                            NetworkCode.SUCCESS -> {
                                viewModel.profileNetworkState.removeObserver(this)

                                App.instance.state.login.loggedIn()

                                hideAnyProgress()
                                activity?.finish()
                            }
                            else                -> handleCode(it.code)
                        }
                    }
                }
            )

        showContent()
    }

    override fun onResume() {
        super.onResume()

        if (UserManager.isAuthorized) {
            activity?.finish()
            return
        }

        token?.let {
            showBlockingProgress()

            val userStorage = UserStorage()
            userStorage.accessToken = it

            viewModel.requestUserWithProfile()
        }
    }

    private fun handleCode(code: NetworkCode) {
        if (code != NetworkCode.STARTED && code != NetworkCode.SUCCESS) {
            UserManager.logout()
            hideAnyProgress()
            when (code) {
                NetworkCode.NO_NETWORK -> showNetworkRequired()
                else                   -> showLoginFailedToast()
            }
        }
    }

    private fun login(view: View) {
        hideKeyboard(view)
        val email = editTextEmail.text.toString().trim { it <= ' ' }
        val password = editTextPassword.text.toString()

        if (!isEmailValid(email)) {
            editTextEmail.error = getString(R.string.error_email)
            return
        }

        if (password.isEmpty()) {
            editTextPassword.error = getString(R.string.error_password)
            return
        }

        showBlockingProgress()
        viewModel.login(email, password)
    }

    private fun startSSOLogin() {
        val intent = LoginHelper.ssoLoginIntent(requireActivity())
        startActivity(intent)
    }

    private fun showLoginFailedToast() {
        showToast(R.string.toast_log_in_failed)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            NavUtils.navigateUpFromSameTask(activity!!)
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    private fun hideKeyboard(view: View) {
        activity?.apply {
            (getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager)
                .hideSoftInputFromWindow(view.windowToken, 0)
        }
    }

    private fun isEmailValid(email: CharSequence): Boolean {
        return android.util.Patterns.EMAIL_ADDRESS.matcher(email).matches()
    }

    override fun onRefresh() {
        hideAnyProgress()
    }

}
